src/handlers/IntentHandler.sol

```diff
@@ -0,0 +1,208 @@
+    tradeOrderHelper = TradeOrderHelper(_tradeOrderHelper);
+    executionFeeInUsd = _executionFeeInUsd;
+    executionFeeTreasury = _executionFeeTreasury;
+  }
+
+  function executeIntent(ExecuteIntentInputs memory inputs) external {
+    if (inputs.accountAndSubAccountIds.length != inputs.cmds.length) revert IntentHandler_BadLength();
+
+    ExecuteIntentVars memory _localVars;
+
+    // Update price to Pyth
+    pyth.updatePriceFeeds(inputs.priceData, inputs.publishTimeData, inputs.minPublishTime, inputs.encodedVaas);
+
+    _localVars.cmdsLength = inputs.accountAndSubAccountIds.length;
+    _localVars.tpTokens = configStorage.getHlpTokens();
+
+    for (uint256 _i; _i < _localVars.cmdsLength; ) {
+      _localVars.mainAccount = address(uint160(inputs.accountAndSubAccountIds[_i].decodeUint(0, 160)));
+      _localVars.subAccountId = uint8(inputs.accountAndSubAccountIds[_i].decodeUint(160, 8));
+      Command _cmd = Command(inputs.cmds[_i].decodeUint(0, 3));
+
+      if (_cmd == Command.ExecuteTradeOrder) {
+        ExecuteTradeOrderVars memory _vars;
+        _vars.marketIndex = inputs.cmds[_i].decodeUint(3, 8);
+        _vars.sizeDelta = inputs.cmds[_i].decodeInt(11, 54) * 1e22;
+        _vars.triggerPrice = inputs.cmds[_i].decodeUint(65, 54) * 1e22;
+        _vars.acceptablePrice = inputs.cmds[_i].decodeUint(119, 54) * 1e22;
+        _vars.triggerAboveThreshold = inputs.cmds[_i].decodeBool(173);
+        _vars.reduceOnly = inputs.cmds[_i].decodeBool(174);
+        _vars.tpToken = _localVars.tpTokens[uint256(inputs.cmds[_i].decodeUint(175, 7))];
+        _vars.createdTimestamp = inputs.cmds[_i].decodeUint(182, 32);
+        _vars.account = _localVars.mainAccount;
+        _vars.subAccountId = _localVars.subAccountId;
+
+        bytes32 key = keccak256(abi.encode(inputs.accountAndSubAccountIds[_i], inputs.cmds[_i]));
+        if (executedIntents[key]) {
+          revert IntentHandler_IntentReplay();
+        }
+
+        _validateSignature(inputs.cmds[_i], inputs.v[_i], inputs.r[_i], inputs.s[_i], _localVars.mainAccount);
+        _executeTradeOrder(_vars);
+        _collectExecutionFeeFromCollateral(_localVars.mainAccount, _localVars.subAccountId);
+
+        executedIntents[key] = true;
+      }
+
+      unchecked {
+        ++_i;
+      }
+    }
+  }
+
+  function _executeTradeOrder(ExecuteTradeOrderVars memory vars) internal {
+    // try executing order
+    try tradeOrderHelper.execute(vars) {
+      // Execution succeeded
+    } catch Error(string memory errMsg) {
+      _handleOrderFail(vars, bytes(errMsg));
+    } catch Panic(uint /*errorCode*/) {
+      _handleOrderFail(vars, bytes("Panic occurred while executing trade order"));
+    } catch (bytes memory errMsg) {
+      _handleOrderFail(vars, errMsg);
+    }
+  }
+
+  function _collectExecutionFeeFromCollateral(address _primaryAccount, uint8 _subAccountId) internal {
+    bytes32[] memory _hlpAssetIds = configStorage.getHlpAssetIds();
+    uint256 _len = _hlpAssetIds.length;
```

Comment:
Instead of loop through all hlpAssetIds, you can loop to `vaultStorage.tradeTokens` instead.